compiler/noirc_frontend/src/elaborator/expressions.rs

```diff
@@ -230,6 +236,41 @@ impl Elaborator<'_> {
         (HirBlockExpression { statements }, block_type)
     }
 
+    /// If the given type was declared as:
+    /// - `#[must_use = "message"]`, return [MustUse::MustUse(Some("message"))]
+    /// - `#[must_use]`, return [MustUse::MustUse(None)]
+    /// - otherwise, return `MustUse::NoMustUse`
+    fn type_is_must_use(typ: &Type) -> MustUse {
+        /// Helper function to avoid infinite recursion for infinitely recursive types
+        fn helper(typ: &Type, fuel: u32) -> MustUse {
+            if fuel == 0 {
+                return MustUse::NoMustUse;
+            }
+            let fuel = fuel - 1;
+            match typ.follow_bindings_shallow().as_ref() {
+                Type::DataType(data_type, _generics) => data_type.borrow().must_use.clone(),
+                // If any element in the tuple is `#[must_use]`, the whole tuple is
+                Type::Tuple(elements) => {
+                    for element in elements {
+                        if let MustUse::MustUse(message) = helper(element, fuel) {
+                            return MustUse::MustUse(message);
+                        }
+                    }
+                    MustUse::NoMustUse
+                }
+                Type::Alias(alias, generics) => helper(&alias.borrow().get_type(generics), fuel),
+                Type::CheckedCast { to, .. } => helper(to.as_ref(), fuel),
+                Type::Reference(element, _) => helper(element.as_ref(), fuel),
+                _ => MustUse::NoMustUse,
```

Comment:
Why don't `Type::Array` and `Type::Slice` inherit the `MustUse` of the item type, similar to how tuples do?